src/app/modules/payments/stripe.service.ts

```diff
@@ -1009,15 +1011,155 @@ export const verifyPaymentStatusWithStripe = (
+ *
+ * @param {IPaymentSchema} payment the payment object, used to form the object path
+ * @param {Buffer} pdfBuffer the pdf to store into s3
+ *
+ * @returns ok(undefined) if no errors are thrown while uploading to s3
+ * @returns err(InvoiceUploadS3Error) if an error is thrown while uploading to s3
+ */
+const _storePaymentProofInS3 = (
+  payment: IPaymentSchema,
+  pdfbuffer: Buffer,
+): ResultAsync<undefined, PaymentProofUploadS3Error> => {
+  const objectPath = getPaymentProofObjectPath(payment)
+
+  logger.info({
+    message: 'Uploading payment proof to s3',
+    meta: {
+      action: '_storePaymentProofInS3',
+      paymentId: payment._id,
+      objectPath,
+      bucket: AwsConfig.paymentProofS3Bucket,
+    },
+  })
+
+  return ResultAsync.fromPromise(
+    AwsConfig.s3
+      .upload({
+        Bucket: AwsConfig.paymentProofS3Bucket,
+        Key: objectPath,
+        Body: Buffer.from(pdfbuffer),
+      })
+      .promise(),
+    (error) => {
+      logger.error({
+        message: 'Error occured whilst uploading pdfBuffer to S3',
+        meta: {
+          action: 'storePaymentProofInS3',
+          paymentId: payment._id,
+          objectPath,
+        },
+        error,
+      })
+      return new PaymentProofUploadS3Error()
+    },
+  )
+    .map(() => {
+      if (payment.completedPayment) {
+        payment.completedPayment = {
+          ...payment.completedPayment,
+          hasS3ReceiptUrl: true,
+        }
+      }
+      return ResultAsync.fromPromise(payment.save(), (error) => {
+        logger.error({
+          message: 'Error occured whilst updating payment',
+          meta: {
+            action: 'storePaymentProofInS3',
+            paymentId: payment._id,
+            objectPath,
+          },
+          error,
+        })
+        return new PaymentProofUploadS3Error()
+      })
+    })
+    .map(() => {
+      return undefined
+    })
+}
+
+/**
+ * Function to generates a presigned url for payment proof stored in s3
+ * Presigned link expires in 30 days
+ *
+ * @param {IPaymentSchema} payment the payment object, used to form the object path
+ *
+ * @returns ok(string) which represents the presigned url if no errors are thrown while generating the presigned url
+ * @returns err(InvoicePresignS3Error) if an error is thrown while generating the presigned link
+ */
+const _getPaymentProofPresignedS3Url = (
+  payment: IPaymentSchema,
+): ResultAsync<string, PaymentProofPresignS3Error> => {
+  const dayInSeconds = 24 * 60 * 60
+  const objectPath = getPaymentProofObjectPath(payment)
+
+  logger.info({
+    message: 'Generating payment proof presigned s3 link',
+    meta: {
+      action: '_getPaymentProofPresignedS3Url',
+      paymentId: payment._id,
+      objectPath,
+    },
+  })
+  return ResultAsync.fromPromise(
+    AwsConfig.s3.getSignedUrlPromise('getObject', {
+      Bucket: AwsConfig.paymentProofS3Bucket,
+      Key: objectPath,
+      Expires: 30 * dayInSeconds,
```

Comment:
/minor suggestion: add a comment here that the presigned link is used for immediate redirect, so the expiry doesn't affect user

Reply:
Left as jsdoc comment